### PR TITLE
114 Progress Stage

### PR DESCRIPTION
**Pull request by andycochran**
This PR… 
- Upgrades Express
- Combines "Approved" and "Withdrawn" project status into "Complete"
- Prevents barfing on null project status query
- Adds default tiles/bounds for null project query
- REQUIRES FRONT END CHANGES (https://github.com/NYCPlanning/labs-zap-search/pull/132)

### CODE REVIEW

File: routes/projects.js [hunk 1]
```diff
@@ -62,6 +61,23 @@ router.get('/', async (req, res) => {
     },
   } = req;
 
+  // if no public status filters are passed use defaults
+  // TODO: Make sure front end doesn't pass a public status query param if no status checkbox is checked
+  if (req.query.dcp_publicstatus.includes('')) req.query.dcp_publicstatus = ['akjdhkjashd', 'alskdjh'];
```
Comment (allthesignals) [inline routes/projects.js]: Huh? What is `akjdhkjashd`?

File: routes/projects.js [hunk 2]
```diff
@@ -208,10 +230,12 @@ router.get('/tiles/:tileId/:z/:x/:y.mvt', async (req, res) => {
...
-  const projectIds = await tileCache.get(tileId);
+  let projectIds = await tileCache.get(tileId);
   // calculate the bounding box for this tile
   const bbox = mercator.bbox(x, y, z, false);
 
+  if (projectIds.length === 0) projectIds = ['\'noprojects\''];
+
   // SELECT data for the vector tile, filtering on the list of projectids
   const SQL = `
     SELECT ST_AsMVT(q, 'project-centroids', 4096, 'geom')
```
Comment (allthesignals) [inline routes/projects.js:242]: Do ya'll think this route is getting too long and hard to understand? Can we split out `/tiles` in a separate file?

File: routes/projects.js [hunk 3]
```diff
@@ -62,6 +61,23 @@ router.get('/', async (req, res) => {
...
+    query: {
+      dcp_publicstatus = ['Approved', 'Withdrawn', 'Filed', 'Certified', 'Unknown'],
+    },
+  } = req;
+
+  if (dcp_publicstatus.includes('Complete')) {
+    dcp_publicstatus.push('Approved');
+    dcp_publicstatus.push('Withdrawn');
+    dcp_publicstatus = dcp_publicstatus.filter(d => d !== 'Complete');
+  }
```
Comment (allthesignals) [inline routes/projects.js:75]: Is there a reason we can't do this in SQL? This seems very difficult to parse through..

**Review by allthesignals — CHANGES_REQUESTED** [review]

File: routes/projects.js (hunk 1 above)
Comment (andycochran) [reply routes/projects.js]: Ha! @chriswhong, we left that in there.

Comment (andycochran) [reply routes/projects.js]: Fixed.

File: routes/projects.js (hunk 2 above)
Comment (andycochran) [reply routes/projects.js:242]: Created #19 for separating these.

File: routes/projects.js (hunk 3 above)
Comment (chriswhong) [reply routes/projects.js:75]: I don't think doing in SQL is the right approach, we are using the nice placeholder so all we need to do is pass in a nice array of values to match against in  `WHERE coalesce(dcp_publicstatus, 'Unknown') IN (${dcp_publicstatus:csv})`.  

We need to substitute "Complete" for "Approved" or "Withdrawn", and I think this is the right place to do it.

Comment (allthesignals) [reply routes/projects.js:75]: What is the "nice placeholder"? I thought customer wants globally "Approved" and "Withdrawn" to = "Complete"?

Comment (chriswhong) [reply routes/projects.js:75]: That's right, but 'Complete' is not a projectstatus in the database, so somewhere we need to tell it what to actually retrieve.

**Pull request merged**